### ISSUE TITLE
sentinel shall have audit logs too on standard out

### ISSUE THREAD

**Issue by v0lkan** [vmware-tanzu/secrets-manager#227]
Sentinel dumps logs into the cli of the caller app but it does not send anything to pod’s std out.

The commands that the user calls, and other important things, including it successful bootstrap shall be sent to the std out as an audit log.

**Comment by v0lkan** [comment]
@farhan-pasha This might pique your interest.

**Comment by sahinakyol** [comment]
Could you assign to me @v0lkan ? If Farhan won't assign to himself.

**Comment by v0lkan** [comment]
@sahinakyol done — we might need to sync up on that — ping me over discord sometime.

**Comment by farhan-pasha** [comment]
hey @sahinakyol, please go ahead with this one. I can see Volkan already assigned this to you.

Thanks
Farhan

**Comment by v0lkan** [comment]
Hey @sahinakyol, I just realized why the approach works on an actual machine but fails on the container. — Our container images are “distroless” and very lightweight for security reasons.

`syscall.Mkfifo()` doesn’t work because there is no `mkfifo` in the system.

Distroless containers are minimalistic containers that contain only the application and its runtime dependencies. They do not include package managers, shells, or any other programs you would expect to find in a standard Linux distribution. The concept behind distroless is to improve security by reducing attack surfaces and to minimize the size of the container.

Unix pipes, which are used to pass output from one program as input to another, are a feature of Unix-like operating systems and are typically used in shell environments. Since distroless containers do not contain a shell or the typical Unix utilities, you cannot use Unix pipes in the traditional sense within a distroless container.

However, if the application running within the distroless container is designed to read from standard input (stdin) and write to standard output (stdout), it can still participate in piping data if it is orchestrated by a container runtime or orchestrator that supports such operations. This would be done outside of the container itself, as the container runtime could handle the piping between containers or between a container and the host system.

So one solution could be to mount the "piped file” as an external volume, but that can be an overkill, and also a security breach (depending on how we mount it)

A different option that come to my mind is to use `gRPC` and create a mini logging gRPC server — well that’s also an overkill—but if I were to pick one poison among the two equally weird ones, I’d choose the one that kills me from the inside (*well bad analogy, but you get it I guess :)*)

I’m pasting a bunch of sandbox code I made; I haven’t tested them at all. So use it with a salt shaker in hand ;) 

```go
syntax = "proto3";

package log;

// The logging service definition.
service Logger {
  // Sends a log message
  rpc LogMessage (LogRequest) returns (LogResponse) {}
}

// The request message containing the log.
message LogRequest {
  string message = 1;
}

// The response message.
message LogResponse {
  string result = 1;
}
```

```bash
protoc --go_out=. --go_opt=paths=source_relative \
    --go-grpc_out=. --go-grpc_opt=paths=source_relative \
    log.proto
```

```go
ype server struct {
    pb.UnimplementedLoggerServer
}

func (s *server) LogMessage(ctx context.Context, in *pb.LogRequest) (*pb.LogResponse, error) {
    log.Printf("Received: %v", in.GetMessage())
    return &pb.LogResponse{Result: "Logged"}, nil
}

func main() {
    lis, err := net.Listen("tcp", ":50051")
    if err != nil {
        log.Fatalf("failed to listen: %v", err)
    }
    s := grpc.NewServer()
    pb.RegisterLoggerServer(s, &server{})
    if err := s.Serve(lis); err != nil {
        log.Fatalf("failed to serve: %v", err)
    }
}
```

```go
package main

import (
    "context"
    "log"
    "google.golang.org/grpc"
    pb "path/to/your/generated/proto/package"
    "time"
)

func main() {
    conn, err := grpc.Dial(":50051", grpc.WithInsecure(), grpc.WithBlock())
    if err != nil {
        log.Fatalf("did not connect: %v", err)
    }
    defer conn.Close()

    c := pb.NewLoggerClient(conn)

    ctx, cancel := context.WithTimeout(context.Background(), time.Second)
    defer cancel()

    r, err := c.LogMessage(ctx, &pb.LogRequest{Message: "Hello, this is a log message"})
    if err != nil {
        log.Fatalf("could not log: %v", err)
    }
    log.Printf("Log Server Response: %s", r.GetResult())
}
```

^ using an approach like this should work in theory :) . Good luck! Say “hi” to yenge.

**Comment by v0lkan** [comment]
@farhan-pasha , I’m not sure if you have had a chance to investigate this issue yet @sahinakyol solution works well as far as I can observe.

There are some failing tests, but that’s more of a testing harness setup problem rather than an issue with the feature (*and eventually, we’ll need to bite the bullet and move away from bash-based testing to a more go-based integration testing—but that’s another story for another time*)

If you have comments and suggestions, feel free to share them.

@sahinakyol, thanks again for implementing this feature; it’s a significant step toward better security compliance and auditabiltiy.

**Comment by v0lkan** [comment]
(p.s. The recent changes have been merged to `main`)
(p.p.s. Tests on `main` are failing and I’m aware of it; I let @sahinakyol know, and he’s gonna amend a fix)

**Comment by farhan-pasha** [comment]
Hey @v0lkan I acknowledge this and plan to test the fix in few hours. I will post an update shortly. Thanks

**Comment by sahinakyol** [comment]
Hey @farhan-pasha and @v0lkan I fixed problem with this [#515](https://github.com/vmware-tanzu/secrets-manager/issues/515) I'm going to create new PR asap. thanks

**Comment by v0lkan** [comment]
Closing this one too. We might want to enhance it later; yet, it’s good enough for now.
